plugins/infinispan52/src/main/java/org/radargun/service/HotRodOperations.java

```diff
@@ -188,5 +205,75 @@ public boolean putIfAbsent(K key, V value, long lifespan, long maxIdleTime) {
             log.tracef("PUT_IF_ABSENT_WITH_LIFESPAN_AND_MAXIDLE cache=%s key=%s value=%s lifespan=%s maxIdle=%s", forceReturn.getName(), key, value, lifespan, maxIdleTime);
          return forceReturn.putIfAbsent(key, value, lifespan, TimeUnit.MILLISECONDS, maxIdleTime, TimeUnit.MILLISECONDS) == null;
       }
+
+   }
+
+   protected class HotRodBulkOperationsCache<K, V> extends HotRodCache<K, V> implements BulkOperations.Cache<K, V> {
+
+      protected final boolean preferAsync;
+
+      public HotRodBulkOperationsCache(RemoteCache<K, V> noReturn, RemoteCache<K, V> forceReturn, boolean preferAsync) {
+         super(noReturn, forceReturn);
+         this.preferAsync = preferAsync;
+      }
+
+      @Override
+      public Map<K, V> getAll(Set<K> keys) {
+         if (trace) log.tracef("GET_ALL cache=%s keys=%s", forceReturn.getName(), keys);
+         Map<K, V> result = new HashMap<>(keys.size());
+         if (preferAsync) {
+            Map<K, Future<V>> futureMap = new HashMap<>(keys.size());
+            for (K key : keys) {
+               futureMap.put(key, forceReturn.getAsync(key));
+            }
+            for (Map.Entry<K, Future<V>> entry : futureMap.entrySet()) {
+               try {
+                  V value = entry.getValue().get();
+                  if (value != null) {
+                     result.put(entry.getKey(), value);
+                  }
+               } catch (Exception e) {
+                  throw new RuntimeException(e);
+               }
+            }
+         } else {
+            for (K key : keys) {
+               result.put(key, forceReturn.get(key));
+            }
+         }
+         return result;
+      }
+
+      @Override
+      public void putAll(Map<K, V> entries) {
+         if (trace) log.tracef("PUT_ALL cache=%s keys=%s", noReturn.getName(), entries);
+         if (preferAsync) {
+            noReturn.putAllAsync(entries);
```

Comment:
Incorrect, this should call noReturn.putAsync in a loop (and then wait for all the futures), not just do the operation asynchronously.